homeassistant/components/bluetooth/__init__.py

```diff
@@ -0,0 +1,267 @@
+        self._cancel_device_detected: CALLBACK_TYPE | None = None
+        self._callbacks: list[tuple[BluetoothCallback, BluetoothMatcher | None]] = []
+
+    async def async_setup(self) -> None:
+        """Set up BT Discovery."""
+        try:
+            self.scanner = HaBleakScanner(
+                scanning_mode=SCANNING_MODE_TO_BLEAK[self.scanning_mode]
+            )
+        except (FileNotFoundError, BleakError) as ex:
+            _LOGGER.warning(
+                "Could not create bluetooth scanner (is bluetooth present and enabled?): %s",
+                ex,
+            )
+            return
+        install_multiple_bleak_catcher(self.scanner)
+        # We have to start it right away as some integrations might
+        # need it straight away.
+        _LOGGER.debug("Starting bluetooth scanner")
+        self.scanner.register_detection_callback(self.scanner.async_callback_dispatcher)
+        self._cancel_device_detected = self.scanner.async_register_callback(
+            self._device_detected, {}
+        )
+        self.hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, self.async_stop)
+        await self.scanner.start()
+
+    @hass_callback
+    def _device_detected(
+        self, device: BLEDevice, advertisement_data: AdvertisementData
+    ) -> None:
+        """Handle a detected device."""
+        matched_domains = {
+            matcher["domain"]
+            for matcher in self._integration_matchers
+            if _ble_device_matches(matcher, device, advertisement_data)
+        }
+        _LOGGER.debug(
+            "Device detected: %s with advertisement_data: %s matched domains: %s",
+            device,
+            advertisement_data,
+            matched_domains,
+        )
+        if not matched_domains and not self._callbacks:
+            return
+
+        service_info: BluetoothServiceInfo | None = None
+        for callback, matcher in self._callbacks:
+            if matcher is None or _ble_device_matches(
+                matcher, device, advertisement_data
+            ):
+                if service_info is None:
+                    service_info = BluetoothServiceInfo.from_advertisement(
+                        device, advertisement_data
+                    )
+                try:
+                    callback(service_info, BluetoothChange.ADVERTISEMENT)
+                except Exception:  # pylint: disable=broad-except
+                    _LOGGER.exception("Error in bluetooth callback")
+
+        if not matched_domains:
+            return
+        if service_info is None:
+            service_info = BluetoothServiceInfo.from_advertisement(
+                device, advertisement_data
+            )
+        for domain in matched_domains:
+            discovery_flow.async_create_flow(
+                self.hass,
+                domain,
+                {"source": config_entries.SOURCE_BLUETOOTH},
+                service_info,
+            )
+
+    async def async_register_callback(
```

Comment:
this doesn't have to be async.

Reply:
e3e1ee525679508a08ef7063d113f6e97a2ecb6f